File: flutter_app/lib/models/templates.dart
class MicrocycleTemplateDto {
  final int? id;
  final String name;
  final String? notes;
  final int orderIndex;
  final int? daysCount;
  final Map<String, dynamic>? schedule;

  MicrocycleTemplateDto({this.id, required this.name, this.notes, this.orderIndex = 0, this.daysCount, this.schedule});

  factory MicrocycleTemplateDto.fromJson(Map<String, dynamic> json) {
    return MicrocycleTemplateDto(
      id: json['id'] as int?,
      name: (json['name'] ?? '').toString(),
      notes: json['notes']?.toString(),
      orderIndex: (json['order_index'] as int?) ?? 0,
      daysCount: json['days_count'] as int?,
      schedule: (json['schedule'] as Map?)?.cast<String, dynamic>(),
    );
  }
}

class MesocycleTemplateResponse {
  final int id;
  final String userId;
  final String name;
  final String? notes;
  final int? weeksCount;
  final int? microcycleLengthDays;
  final double? normalizationValue;
  final String? normalizationUnit;
  final bool isPublic;
  final List<MicrocycleTemplateDto> microcycles;

  MesocycleTemplateResponse({
    required this.id,
    required this.userId,
    required this.name,
    this.notes,
    this.weeksCount,
    this.microcycleLengthDays,
    this.normalizationValue,
    this.normalizationUnit,
    required this.isPublic,
    required this.microcycles,
  });

  factory MesocycleTemplateResponse.fromJson(Map<String, dynamic> json) {
    return MesocycleTemplateResponse(
      id: json['id'] as int,
      userId: (json['user_id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      notes: json['notes']?.toString(),
      weeksCount: json['weeks_count'] as int?,
      microcycleLengthDays: json['microcycle_length_days'] as int?,
      normalizationValue: (json['normalization_value'] as num?)?.toDouble(),
      normalizationUnit: json['normalization_unit']?.toString(),
      isPublic: (json['is_public'] as bool?) ?? false,
      microcycles: (json['microcycles'] as List<dynamic>? ?? [])
          .whereType<Map<String, dynamic>>()
          .map((e) => MicrocycleTemplateDto.fromJson(e))
          .toList(),
    );
  }
}

File: services/plans-service/plans_service/schemas/templates.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field, constr


class MicrocycleTemplateDto(BaseModel):
    id: Optional[int] = None
    name: str = Field(..., max_length=255)
    notes: Optional[constr(max_length=255)] = None
    order_index: int = Field(default=0, ge=0)
    days_count: Optional[int] = Field(default=None, ge=1, le=31)
    schedule: Optional[Dict[str, List[Dict[str, Any]]]] = None

    class Config:
        from_attributes = True


class PlacementMode(str):
    Insert_After_Mesocycle = "Insert_After_Mesocycle"
    Insert_After_Workout = "Insert_After_Workout"
    Append_To_End = "Append_To_End"


class ConflictPolicy(str):
    Shift_Forward = "Shift_Forward"
    Replace_Planned = "Replace_Planned"
    Skip_On_Conflict = "Skip_On_Conflict"


class PlacementDto(BaseModel):
    mode: str = Field(
        ...,
        description="Placement mode: Insert_After_Mesocycle | Insert_After_Workout | Append_To_End",
    )
    mesocycle_index: Optional[int] = Field(default=None, ge=1)
    workout_id: Optional[int] = Field(default=None, ge=1)


class InstantiateFromTemplateRequest(BaseModel):
    template_id: int
    placement: PlacementDto
    on_conflict: str = Field(default=ConflictPolicy.Shift_Forward)


class InstantiateFromExistingRequest(BaseModel):
    source_mesocycle_id: int
    placement: PlacementDto
    on_conflict: str = Field(default=ConflictPolicy.Shift_Forward)


class MesocycleTemplateCreate(BaseModel):
    name: str = Field(..., max_length=255)
    notes: Optional[constr(max_length=255)] = None
    weeks_count: Optional[int] = Field(default=None, ge=0)
    microcycle_length_days: Optional[int] = Field(default=None, ge=1, le=14)
    normalization_value: Optional[float] = Field(default=None)
    normalization_unit: Optional[str] = Field(default=None)
    is_public: bool = Field(default=False)
    microcycles: List[MicrocycleTemplateDto] = Field(default_factory=list)


class MesocycleTemplateUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=255)
    notes: Optional[constr(max_length=255)] = None
    weeks_count: Optional[int] = Field(default=None, ge=0)
    microcycle_length_days: Optional[int] = Field(default=None, ge=1, le=14)
    normalization_value: Optional[float] = Field(default=None)
    normalization_unit: Optional[str] = Field(default=None)
    is_public: Optional[bool] = None
    microcycles: Optional[List[MicrocycleTemplateDto]] = None


class MesocycleTemplateResponse(BaseModel):
    id: int
    user_id: str
    name: str
    notes: Optional[str]
    weeks_count: Optional[int]
    microcycle_length_days: Optional[int]
    normalization_value: Optional[float]
    normalization_unit: Optional[str]
    is_public: bool
    microcycles: List[MicrocycleTemplateDto] = Field(default_factory=list)

    class Config:
        from_attributes = True
